plugins/GSdx/GSFastList.h

```diff
@@ -0,0 +1,252 @@
+/*
+*	Copyright (C) 2017-2017 Alessandro Vetere
+*	Copyright (C) 2007-2009 Gabest
+*	http://www.gabest.org
+*
+*  This Program is free software; you can redistribute it and/or modify
+*  it under the terms of the GNU General Public License as published by
+*  the Free Software Foundation; either version 2, or (at your option)
+*  any later version.
+*
+*  This Program is distributed in the hope that it will be useful,
+*  but WITHOUT ANY WARRANTY; without even the implied warranty of
+*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+*  GNU General Public License for more details.
+*
+*  You should have received a copy of the GNU General Public License
+*  along with GNU Make; see the file COPYING.  If not, write to
+*  the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA USA.
+*  http://www.gnu.org/copyleft/gpl.html
+*
+*/
+
+#pragma once
+
+template <class T>
+struct Element {
+	T  data;
+	uint16 next_index;
+	uint16 prev_index;
+};
+
+template <class T>
+class FastListIterator;
+
+// T should be either a primitive type or a pointer
+template <class T>
+class FastList {
+	friend class FastListIterator<T>;
+private:
+	// The index of the first element of the list is m_buffer[0].next_index
+	//     The first Element<T> of the list has prev_index equal to 0
+	// The index of the last element of the list is m_buffer[0].prev_index
+	//     The last Element<T> of the list has next_index equal to 0
+	// All the other Element<T> of the list are chained by next_index and prev_index
+	// m_buffer has dynamic size m_capacity
+	// m_buffer[0] is always present as auxiliary Element<T> of the list
+	Element<T>* m_buffer;
+	uint16 m_capacity;
+	uint16 m_free_indexes_stack_top;
+	// m_free_indexes_stack has dynamic size (m_capacity - 1)
+	// m_buffer indexes that are free to be used are stacked here
+	uint16* m_free_indexes_stack;
+
+public:
+	__forceinline FastList() {
+		m_buffer = nullptr;
+		clear();
+	}
+
+	__forceinline ~FastList() {
+		_aligned_free(m_buffer);
+	}
+
+	void clear() {
+		// Initialize m_capacity to 4 so we avoid to Grow() on initial insertions
+		// The code doesn't break if this value is changed with anything from 1 to USHRT_MAX
+		m_capacity = 4;
+		// Initialize m_buffer and m_free_indexes_stack as a contiguous block of memory starting at m_buffer
+		// This should increase cache locality and reduce memory fragmentation
+		_aligned_free(m_buffer);
+		m_buffer = (Element<T>*)_aligned_malloc(m_capacity * sizeof(Element<T>) + (m_capacity - 1) * sizeof(uint16), 64);
+		m_free_indexes_stack = (uint16*)&m_buffer[m_capacity];
+		// Initialize m_buffer[0]
+		m_buffer[0].data = nullptr; // Unused field
```

Comment:
It is `m_buffer[0] = {nullptr, 0, 0};`, but I agree it is nicer. Will change...maybe, read below please.
That's the occasion to discuss a problem here: `m_buffer[0].data` is meant to be unused, but initializing it with `nullptr` makes the code non compiling for every `T` which is not a pointer. Should I drop it's initialization, so to have compilation for every type? Beware that some design choice of `FastList`, that is returning `T` instead of `T&` when accessing the data, makes it unusable for every non primitive-non pointer type.